ark/phenotyping/som_utils_test.py

```diff
@@ -11,115 +11,285 @@
 
-    # we will assign cluster labels based on the average of the rows
-    sample_pixel_mat['Cluster_ID'] = sample_pixel_mat.mean(axis=1)
+    # multiply by 100 and truncate to int to get an actual cluster id
+    cluster_ids = sub_means * 100
+    cluster_ids = cluster_ids.astype(int)
 
-    # multiply by 100 and truncate for an integer cluster assignment
-    sample_pixel_mat['Cluster_ID'] *= 100
-    sample_pixel_mat['Cluster_ID'] = sample_pixel_mat['Cluster_ID'].astype(int)
+    # we will now broadcast these cluster_ids across each DataFrame in pixel_mat_preprocessed.hdf5
+    for fov in fovs:
+        # get the fovs pixel matrix
+        flowsom_clustered = pd.read_hdf(os.path.join(base_dir, pixel_pre_name), key=fov)
 
-    # write the final data to CSV
-    sample_pixel_mat.to_csv(os.path.join(base_dir, 'pixel_mat_clustered.csv'), index=False)
+        # repeat the same cluster id 10 times for each submatrix
+        flowsom_clustered['Cluster'] = cluster_ids.repeat(5).reset_index(drop=True)
+
+        # write clustered data to HDF5
+        flowsom_clustered.to_hdf(os.path.join(base_dir, pixel_cluster_name), key=fov, mode='a')
 
 
 def test_create_pixel_matrix():
+    fovs = ['fov0', 'fov1']
+    chans = ['chan0', 'chan1']
+
     # must be float64, since that's how Candace's data comes in
     sample_img_xr = test_utils.make_images_xarray(tif_data=None,
-                                                  fov_ids=['fov0', 'fov1'],
-                                                  channel_names=['chan0', 'chan1'],
+                                                  fov_ids=fovs,
+                                                  channel_names=chans,
                                                   dtype='float64')
 
     sample_labels = test_utils.make_labels_xarray(label_data=None,
-                                                  fov_ids=['fov0', 'fov1'],
+                                                  fov_ids=fovs,
                                                   compartment_names=['whole_cell'])
 
-    # pass invalid fov names
-    with pytest.raises(ValueError):
-        som_utils.create_pixel_matrix(sample_img_xr, sample_labels, fovs=['fov1', 'fov2'])
+    # basic error checking
+    with tempfile.TemporaryDirectory() as temp_dir:
+        # pass invalid fov names
+        with pytest.raises(ValueError):
+            som_utils.create_pixel_matrix(sample_img_xr, sample_labels, temp_dir,
+                                          fovs=['fov1', 'fov2'])
+
+        # pass invalid channel names
+        with pytest.raises(ValueError):
+            som_utils.create_pixel_matrix(sample_img_xr, sample_labels, temp_dir,
+                                          channels=['chan1', 'chan2'])
+
+        # pass invalid base directory
+        with pytest.raises(FileNotFoundError):
+            som_utils.create_pixel_matrix(sample_img_xr, sample_labels, 'bad_base_dir')
+
+    # test on all fovs and channels, use default file name
```

Comment:
Can you turn the tests below into two nested for loops, which loop over the fovs and channels which are passed to the function? That way we don't need to rewrite all of the code for each test, it'll just use the supplied channel/fov list